rl_studio/agents/robot_mesh/manual_pilot.py

```diff
@@ -0,0 +1,130 @@
+        self.env = gym.make(self.env_name, **env_params)
+        # algorithm params
+        self.alpha = params.algorithm["params"]["alpha"]
+        self.epsilon = params.algorithm["params"]["epsilon"]
+        self.gamma = params.algorithm["params"]["gamma"]
+        # agent
+        # self.action_number = params.agent["params"]["actions_number"]
+        # self.actions_set = params.agent["params"]["actions_set"]
+        # self.actions_values = params.agent["params"]["available_actions"][self.actions_set]
+
+    def main(self):
+
+        print(f"\t- Start hour: {datetime.datetime.now()}\n")
+        print(f"\t- Environment params:\n{self.environment_params}")
+        config = QLearnConfig()
+
+        # TODO: Move to settings file
+        outdir = "./logs/robot_mesh_experiments/"
+        stats = {}  # epoch: steps
+        states_counter = {}
+        states_reward = {}
+
+        last_time_steps = np.ndarray(0)
+
+        env = gym.wrappers.Monitor(self.env, outdir, force=True)
+        actions = range(env.action_space.n)
+
+        print(f"\t- Start hour: {datetime.datetime.now()}")
+
+        # TODO: Move to settings file
+        outdir = "./logs/robot_mesh_experiments/"
+        stats = {}  # epoch: steps
+        states_counter = {}
+        states_reward = {}
+
+        # plotter = liveplot.LivePlot(outdir)
+
+        last_time_steps = np.ndarray(0)
+
+        actions = range(env.action_space.n)
+        env = gym.wrappers.Monitor(self.env, outdir, force=True)
+        counter = 0
+        lap_completed = False
+        total_episodes = 20000
+        epsilon_discount = 0.999  # Default 0.9986
+        rewards_per_run = [0, 0]
+        env.done = False
+
+        # START ############################################################################################################
+        for episode in range(total_episodes):
+
+            counter = 0
+            lap_completed = False
+
+            initial_epsilon = 0.999
+            highest_reward = 0
+            cumoviulated_reward = 0
+            print("resetting")
+            time.sleep(5)
+            state = env.reset()
+
+            # state = ''.join(map(str, observation))
+
+            for step in range(50000):
+
+                inpt = input("provide action (0-up, 1-right, 2-down, 3-left): ")
+                if inpt == "0" or inpt == "1" or inpt == "2" or inpt == "3":
+                    action = int(inpt)
+                    print("Selected Action!! " + str(action))
+                    # Execute the action and get feedback
+                    nextState, reward, env.done, lap_completed = env.step(action)
+
+                    env._flush(force=True)
+                    if not env.done:
+                        state = nextState
+                    else:
+                        break
+                elif inpt == "q":
+                    system.exit(1)
+                else:
+                    print("wrong action! Try again")
+
```

Comment:
maybe we could break this function into smaller, more manageable functions and take the opportunity to refactor this program, don't you think?

Reply:
Totally right